build: Add verify-release script

jku opened the pull request:
verify-release
* Builds a release from current commit
* Warns if git describe does not match built version (in other words we're building from a commit that is not release tagged)
* Warns if built version is not the latest GitHub or PyPI version
* Errors if GitHub or PyPI release artifacts for this version do not match the built
  release artifacts

This should be useful after release as any developer (or a CI job) can
easily verify that the release matches the sources in git.

Note that the last checks currently fail as the 1.0 build was not
reproducible. They should succeed after next release.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

---


Some comments:
* this commit adds a file to root directory, I don't like it... but I didn't know where else to put it
* signatures are not checked because it's so painful. This could be a future improvement if someone has good ideas...
* if the release files do not match, there is an error but no details are printed and the downloaded files are not left available: I did not want to add this complexity to the script

Commit:
build: Add verify-release script

verify-release
* Builds a release from current commit
* Notifies if git describe does not match built version
* Notifies if built version is not the latest GitHub or PyPI version
* Asserts that the GitHub and PyPI release artifacts match the built
  release artifacts

This should be useful after release as any developer (or a CI job) can
easily verify that the release matches the sources in git.

Note that the last checks currently fail as the 1.0 build was not
reproducible. They should succeed after next release.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented on verify_release:
```diff
@@ -0,0 +1,159 @@
+        subprocess.run(binary_download, stdout=subprocess.DEVNULL, check=True)
+        subprocess.run(source_download, stdout=subprocess.DEVNULL, check=True)
+
+        return not dircmp(pypi_dir, compare_dir).diff_files
```
maybe this should check common_files instead of diff_files to also cover things like a completely missing  release artifact...

jku replied:
yes, this is a bug: if files are only found in one of the directories, then diff_files is empty. So should rather test that `common_files` list contains the two expected filenames.

lukpueh replied:
Are you going to change it (here and above)? Also, did you verify that `common_files` recurses into subdirs?

jku replied:
I'll fix this yes.

jku replied:
the correct field is actually `same_files`. I hard coded the expected list of files (1 tarball, 1 wheel) that must match when comparing releases.

lukpueh commented on verify_release:
```diff
@@ -0,0 +1,159 @@
+def get_pypi_version() -> str:
+    """Return latest version string available on PyPI"""
+    # pip is sad: it can't even tell what the newest available version is...
+    # FIXME: Maybe think of something better than downloading latest tarball here?
```
You could use the [PyPI JSON API project endpoint](https://warehouse.pypa.io/api-reference/json.html#project): `https://pypi.org/pypi/tuf/json`

jku replied:
True. I intentionally wanted to find the version that pip considers current since that's what matters in the end (obviously the two should be the same but...).

I suppose the method name is a bit misleading in that sense

lukpueh replied:
Did you consider using experimental `pip index versions tuf` or parsing error output of `pip install tuf==`? (quoting [SO](https://stackoverflow.com/questions/4888027/python-and-pip-list-all-versions-of-a-package-thats-available/26664162#26664162))

jku replied:
well, I have not found a way to find the answer to _"what is **current** version according to pip?"_ I think none of those answer this question.

lukpueh replied:
Okay, that all makes sense. :)

jku replied:
maybe I've left out the reasoning for this: A malicious maintainer could upload tuf version 99. In this case it's not enough that we check that locally built 1.0 matches pypi version 1.0: end users are still getting version 99 if they do `pip install tuf`.

jku replied:
I've improved naming and comments, did not modify the code

lukpueh reviewed (commented):
Cool stuff, @jku! I tried it out and it works nicely.

Are we supposed to run it with development requirements installed? It only needs `requests` and `build` to be installed. Maybe it's worth mentioning requirements somewhere, e.g. in the doc header or in RELEASE.md?

Will get back with an actual code review.

Commit:
verify_release: Be explicit about PyPI version

We are interested in what pip thinks is the current tuf version: make
that explicit in method naming and comments.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

lukpueh reviewed (changes requested):
LGTM, modulo the bug you discovered yourself.

Commit:
verify_release: Be specific about expected artifacts

Use a hard-coded list of artifacts that we expect to find in a
release. Specifically check that each of those files matches
the corresponding file in locally built release.

Also add two missing annotations.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
Issue is fixed. The testing is obviously not quite 100.0% (since we dont have a positive test case on pypi.org yet) but I have done the manual testing I could.

lukpueh reviewed (approved):
🎉
